### PR TITLE
Allow regex groups *after* keyword groups

### LINKED ISSUE #334: Question: shouldn't regex import groups take precedence over group keywords?
Hey team, thanks a ton for the awesome import magic in here! An issue tho…

#### System information

- OS: OSX El Capitan
- TypeScript Hero version: 1.7.0
- VSCode version: 1.17.2

#### Description

I’ve got the following configuration set up:

```js
{
  // …
  "typescriptHero.resolver.importGroups": [
    "Modules",
    "/material-ui/",
    "Plains",
    "Workspace"
  ]
}
```

The idea is that I'd like my npm-installed imports from material-ui to be in their own groups, below regular modules and plains, and before my workspace files. Say, I should get the following:

```js
import autobind from 'autobind-decorator'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import DocumentTitle from 'react-document-title'
import { connect } from 'react-redux'

import { Card, CardActions, CardText, CardTitle } from 'material-ui/Card'
import RaisedButton from 'material-ui/RaisedButton'
import Snackbar from 'material-ui/Snackbar'
import ArrowForward from 'material-ui/svg-icons/navigation/arrow-forward'
import TextField from 'material-ui/TextField'

import '../styles/LoginScreen.styl'

import { LoginStatePropType } from '../prop-types'
import { logIn } from '../reducers/currentUser'
```

However, material-ui imports get lumped in with the Modules group: their group definition is entirely ignored. Like this:

```js
import autobind from 'autobind-decorator'
import { Card, CardActions, CardText, CardTitle } from 'material-ui/Card'
import RaisedButton from 'material-ui/RaisedButton'
import Snackbar from 'material-ui/Snackbar'
import ArrowForward from 'material-ui/svg-icons/navigation/arrow-forward'
import TextField from 'material-ui/TextField'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import DocumentTitle from 'react-document-title'
import { connect } from 'react-redux'

import '../styles/LoginScreen.styl'

import { LoginStatePropType } from '../prop-types'
import { logIn } from '../reducers/currentUser'
```

Is that as intended?

##### Steps to reproduce

1. Create a workspace
2. Plug the related config fragment in your workspace settings
3. Create a file with the last import list
4. Save (and if not auto-organizing, organize)

Best,

### PR DESCRIPTION
**Pull request by tdd**
It was impossible to define regex groups when keyword groups were listed ahead of them and matching the same source. See #334 for details and examples. This addresses the issue.

- Fixes #334

#### Description

The code change itself is fairly minimal, adjusting the logic of `ImportManager#addImportsToGroups()` to rely on a new utility function. Both have added tests that pass.

#### A word about tests

Despite the badge saying `develop` passes build, it failed on my machine right from the forked code. I get a timeout and 4 undefined import declarations. I don't know enough about TS' internals to figure out why this was, so I left these failing tests as-is, and just added my (passing) ones. No previously-OK test broke in the process, obviously 😉 

Also, there was an implicit assumption by many tests that indent size was 4 (VSCode's default, and the original author's). But anyone running tests with a different User Setting would see a lot of failures. I fixed this by forcing a 4-space indent in the test workspace's settings.

#### VSCode, TypeScript and me…

This is the first time ever I write TypeScript for actual production code (I'm more an ES.Next kinda guy) or write VSCode extension stuff, so I'm sorry if some of my code, test or otherwise, seems non-idiomatic: feel free to point me in the right direction for you. I tried to abide the original coding style (had to disable my ESLint/Prettier settings to avoid unduly reformatting everything), I hope it's good with you!

I do intend to contribute two other PRs to this ext, though: one to address a personal need that I apparently share with many (e.g. #316): an option that lets us sort imports by first specifier name, not by module name. And another purely for "regex grooming": I know JS regexes (and regexes in general) intimately, and it seems to me a lot can be improved on the way they're used by the ext; because this is not feature-related and quite subjective, I'll make a separate PR out of it.

Looking forward to your feedback! (and explanations about why these other tests failed, I'm curious).

Best,

### CODE REVIEW

**Comment by tdd** [comment]
It's odd how I don't get the errors reported by CI on my side. But I did see the causes and pushed fixes, so it should end up working! Very curious why I get *other* failures locally.

File: src/extension/utilities/utilityFunctions.ts [hunk 1]
```diff
@@ -42,6 +43,24 @@ export function stringSort(strA: string, strB: string, order: 'asc' | 'desc' = '
...
+ * keyword groups yet capture relevant imports nonetheless.
+ *
+ * @export
+ * @param {ImportGroup[]} importGroups The original import groups (as per extension configuration)
+ * @returns {ImportGroup[]} The same list, with Regex import groups appearing first.
+ */
+export function importGroupSortForPrecedence(importGroups: ImportGroup[]): ImportGroup[] {
+    const regexGroups: ImportGroup[] = [];
+    const otherGroups: ImportGroup[] = [];
+    for (const ig of importGroups) {
```
Comment (buehler) [inline src/extension/utilities/utilityFunctions.ts]: Actually.. clever :-)

File: test/single-workspace-tests/extension/utilities/utilityFunctions.test.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import * as chai from 'chai';
```
Comment (buehler) [inline test/single-workspace-tests/extension/utilities/utilityFunctions.test.ts]: please switch to an intend of 4 spaces (to adhere with the rest of the code)
(in the file)

**Comment by buehler** [comment]
Hey @tdd 
First of all: thank you! 🎉 

It's a pretty nice code you wrote, so it's all nice and square. Just one little comment to come bye since the rest of the code is written that way.

Did you still experience other errors locally? I tried on my computer @ home and there I don't have any errors (anymore).

**Comment by buehler** [comment]
A question on my side: Do you think this "feature" should be configurable? Or rather not?

File: test/single-workspace-tests/extension/utilities/utilityFunctions.test.ts (hunk 2 above)
Comment (tdd) [reply test/single-workspace-tests/extension/utilities/utilityFunctions.test.ts]: Dang! I did in the other PR (there's gonna be a conflict, btw, as this file is introduced in both, so once #336 is merged in, I'll re-work this so it merges fine).

**Comment by tdd** [comment]
Hey @buehler,

#### Config: nope

I don't think this requires configuration, as the ordering you'd use in `importGroups` *is* the configuration already. I fail to see any use-case in which you'd define regex groups but would want broader keyword groups to preempt them.

#### Local errors: yup 😢 

As for errors: yes, I get them all the time. I wonder if this is due to a different TS version (yarn's lock promise being faulty, it put me with the latest TS 2.6, when the original code was 2.5). I tried disabling all my exts, still got issues (I have to disable Prettier and ESLint anyway when working on this, as my settings for them conflict with either coding or testing it).

This is super-annoying, and I have no idea what gives.

**Pull request merged**